lib/dal/src/migrations/U2424__optimize_component_creation.sql

```diff
@@ -0,0 +1,365 @@
+        FROM attribute_prototypes_v1(this_tenancy, this_visibility) AS ap
+        WHERE in_attribute_context_v1(this_attribute_context, ap);
+
+        SELECT av.object
+        INTO this_new_attribute_value
+        FROM attribute_value_create_v1(
+            this_tenancy,
+            this_visibility,
+            this_attribute_context,
+            this_unset_func_binding_id,
+            this_unset_func_binding_return_value_id,
+            NULL
+        ) AS av;
+
+        PERFORM set_belongs_to_v1(
+            'attribute_value_belongs_to_attribute_prototype',
+            this_tenancy,
+            this_visibility,
+            this_new_attribute_value ->> 'id',
+            this_attribute_prototype.id
+        );
+    END LOOP;
+
+    -- Explicit Internal Providers
+    FOR this_internal_provider IN
+        SELECT *
+        FROM internal_providers_v1(this_tenancy, this_visibility)
+        WHERE schema_variant_id = this_schema_variant_id
+    LOOP
+        this_attribute_context := attribute_context_build_from_parts_v1(
+            ident_nil_v1(), -- Prop ID
+            this_internal_provider.id, -- Internal Provider ID
+            ident_nil_v1(), -- External Provider ID
+            this_new_row.id -- Component ID
+        );
+
+        SELECT *
+        INTO STRICT this_attribute_prototype
+        FROM attribute_prototypes_v1(this_tenancy, this_visibility) AS ap
+        WHERE in_attribute_context_v1(this_attribute_context, ap);
+
+        SELECT av.object
+        INTO this_new_attribute_value
+        FROM attribute_value_create_v1(
+            this_tenancy,
+            this_visibility,
+            this_attribute_context,
+            this_unset_func_binding_id,
+            this_unset_func_binding_return_value_id,
+            NULL
+        ) AS av;
+
+        PERFORM set_belongs_to_v1(
+            'attribute_value_belongs_to_attribute_prototype',
+            this_tenancy,
+            this_visibility,
+            this_new_attribute_value ->> 'id',
+            this_attribute_prototype.id
+        );
+    END LOOP;
+
+    SELECT array_agg(ip.*)
+    INTO STRICT this_internal_providers
+    FROM internal_providers_v1(this_tenancy, this_visibility) AS ip
+        INNER JOIN props_v1(this_tenancy, this_visibility) AS props
+          ON ip.prop_id = props.id
+    WHERE props.schema_variant_id = this_schema_variant_id;
+
+    SELECT array_agg(ap.*)
+    INTO STRICT this_attribute_prototypes
+    FROM attribute_prototypes_v1(this_tenancy, this_visibility) AS ap
+    WHERE ap.attribute_context_internal_provider_id = ANY (SELECT id FROM UNNEST(this_internal_providers))
+          OR ap.attribute_context_prop_id = ANY (SELECT prop_id FROM UNNEST(this_internal_providers));
+
+    SELECT array_agg(av.*)
+    INTO STRICT this_attribute_values
+    FROM attribute_values_v1(this_tenancy, this_visibility) AS av
+    WHERE av.attribute_context_prop_id = ANY (SELECT prop_id FROM UNNEST(this_internal_providers));
```

Comment:
Was there a specific reason to not use the `belongs_to` relationship between `AttributeValue`s and `AttributePrototype`s? There _shouldn't_ be stray `AttributeValue`s sitting around that might match based on the `attribute_context_prop_id`, but I'm pretty sure there are places that aren't cleaning up after themselves as well as they should be (which means there are probably more that I'm not thinking of).
```suggestion
    SELECT array_agg(av.*)
    INTO STRICT this_attribute_values
    FROM attribute_values_v1(this_tenancy, this_visibility) AS av
        INNER JOIN attribute_value_belongs_to_attribute_prototype_v1(this_tenancy, this_visibility) AS avbtap
            ON avbtap.object_id = av.id
    WHERE avbtap.belongs_to_id = ANY (SELECT id FROM UNNEST(this_attribute_prototypes));
```

Reply:
I just copied the query we used inside of the loop to outside of it.

Inside of the loop we used to do this:

```
        SELECT *
        INTO STRICT this_prop_attribute_value
        FROM attribute_values_v1(this_tenancy, this_visibility) AS av
        WHERE in_attribute_context_v1(
            attribute_context_build_from_parts_v1(
                this_internal_provider.prop_id,
                ident_nil_v1(),
                ident_nil_v1(),
                ident_nil_v1()
            ),
            av
        );
```

Since I had the prop ids beforehand I fetched all avs with that prop_id, is my thinking incorrect?